### ISSUE TITLE
 "k8s.io/kubernetes/pkg/scheduler/core".ScheduleAlgorithm has no field or method Snapshot

### ISSUE THREAD

**Issue by lsytj0413** [kubernetes/autoscaler#2858]
* branch
cluster-autoscaler-release-1.17

* go
go version go1.13.6 darwin/amd64

I have a project import autoscaler, it's version is:

```
* go get k8s.io/autoscaler/cluster-autoscaler@cluster-autoscaler-release-1.17

require (
k8s.io/autoscaler/cluster-autoscaler v0.0.0-20200220022703-fbade1d48794
)

replace (
	k8s.io/api => k8s.io/api v0.17.3
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.17.3
	k8s.io/apimachinery => k8s.io/apimachinery v0.17.3
	k8s.io/apiserver => k8s.io/apiserver v0.17.3
	k8s.io/cli-runtime => k8s.io/cli-runtime v0.17.3
	k8s.io/client-go => k8s.io/client-go v0.17.3
	k8s.io/cloud-provider => k8s.io/cloud-provider v0.17.3
	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.17.3
	k8s.io/code-generator => k8s.io/code-generator v0.17.3
	k8s.io/component-base => k8s.io/component-base v0.17.3
	k8s.io/cri-api => k8s.io/cri-api v0.17.3
	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.17.3
	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.17.3
	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.17.3
	k8s.io/kube-proxy => k8s.io/kube-proxy v0.17.3
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.17.3
	k8s.io/kubectl => k8s.io/kubectl v0.17.3
	k8s.io/kubelet => k8s.io/kubelet v0.17.3
	k8s.io/kubernetes => k8s.io/kubernetes v1.17.3
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.17.3
	k8s.io/metrics => k8s.io/metrics v0.17.3
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.17.3
)
```

It failed with:

```
pkg/mod/k8s.io/autoscaler/cluster-autoscaler@v0.0.0-20200220022703-fbade1d48794/simulator/predicates.go:209:31: p.scheduler.Algorithm.Snapshot undefined (type "k8s.io/kubernetes/pkg/scheduler/core".ScheduleAlgorithm has no field or method Snapshot)
```

I found this function is introduce in https://github.com/kubernetes/autoscaler/commit/b67854e8008ffc1f0ead9cd00dbaa12fb08b5efc#diff-aa62a4a46f434572b43666b621f733ef ，but the **Snapshot** function only available at kubernetes release 1.18 (or higher) https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/scheduler/core/generic_scheduler.go#L114 ，doesn't available at kubernetes release 1.17 https://github.com/kubernetes/kubernetes/blob/release-1.17/pkg/scheduler/core/generic_scheduler.go#L114 .

is there something wrong that the autoscaler release-1.17 branch depend on a function only availale at kuberneres release-1.18 (or higher)?

**Comment by MaciekPytel** [comment]
The reason for this is that with changes in scheduler in 1.17, Cluster Autoscaler could no longer work. Adding the Snapshot function was necessary to fix it. Unfortunately we caught the problem a bit too late in development cycle and the fix in scheduler didn't quite make it before the cut. So we vendored a snapshot of Kubernetes shortly after 1.17 was cut instead (knowing that there was no other changes in scheduler and so the version vendored in CA will behave the same as real 1.17 scheduler).

**Comment by lsytj0413** [comment]
> The reason for this is that with changes in scheduler in 1.17, Cluster Autoscaler could no longer work. Adding the Snapshot function was necessary to fix it. Unfortunately we caught the problem a bit too late in development cycle and the fix in scheduler didn't quite make it before the cut. So we vendored a snapshot of Kubernetes shortly after 1.17 was cut instead (knowing that there was no other changes in scheduler and so the version vendored in CA will behave the same as real 1.17 scheduler).

Thanks for your reply. As you mentioned, currently the 1.17 scheduler doesn't have this fix, but there's a plan it will cherry-pick to 1.17？or in my scenario, i must change the kubernetes pkg to 1.18？

**Comment by lsytj0413** [comment]
By the way, I notice that kubernetes remove the ScheduleAlgorithm.Snapshot function in this [Support multiple scheduling profiles in a single scheduler](https://github.com/kubernetes/kubernetes/commit/c0488584712fe37372f35236d2d6dd0e9f630e26#diff-c237cdd9e4cb201118ca380732d7f361L117) commit，from tag v1.18.0-beta.0 to v1.18.0-beta.1 change. Is autoscaler will remove the Snapshot function usage？

**Comment by MaciekPytel** [comment]
> Is autoscaler will remove the Snapshot function usage？

Yes, in 1.18 we're migrating to scheduler framework, which completely changes how CA interacts with scheduler code. As part of it we have implemented our own snapshot implementation and we're no longer going to use Snapshot() function.

**Comment by lsytj0413** [comment]
> > Is autoscaler will remove the Snapshot function usage？
> 
> Yes, in 1.18 we're migrating to scheduler framework, which completely changes how CA interacts with scheduler code. As part of it we have implemented our own snapshot implementation and we're no longer going to use Snapshot() function.

How about CA release-1.17？

**Comment by RainbowMango** [comment]
@MaciekPytel 

Is there any plan to improve the vendor thing? I mean, to support go module. And I can help with it.
Kubernetes has released v1.19, Is it a good time to start?
